# rules for the rmats turbo snakemake pipeline

RMATS_CMD = "python rmats.py "
if CONTAINER_TYPE == "singularity":
    sif_path = os.path.join(TEMP_DIR, "rmats-turbo.sif")
    RMATS_CMD = f"singularity run {sif_path} "
elif CONTAINER_TYPE == "docker":
    sys.exit("docker is unsupported at this point")

def get_rmats_input_config(wildcards):
    " return list of config input files for a given comparison"
    try:
        comp_samples = COMPARISONS[wildcards.comparison]
    except KeyError:
        sys.exit(f"{wildcards.comparison} not found in COMPARISONS config")
    
    if len(comp_samples) != 2:
        sys.exit("COMPARISIONS must contain 2 SAMPLES to compare")
    try:
        res = {
          "b1" : SAMPLE_CONFIGS[comp_samples[0]],
          "b2" : SAMPLE_CONFIGS[comp_samples[1]]
          }
    except:
        smpls = " ".join(comp_samples)
        sys.exit(f"issue finding configuration for samples {smpls}")
    
    return res

rule rmats:
    input:
        unpack(get_rmats_input_config) 
    output:
        xlsx = "{results}/{comparison}/summary.txt"
    params:
        job_name = "rmats_{comparison}",
        memory ="select[mem>50] rusage[mem=50]",
        od = "{results}/{comparison}",
        other_args = OPT_RMATS_ARGS,
        gtf = GTF_PATH,
        read_len = READ_LENGTH,
        tmp = TEMP_DIR,
        star_index = STAR_INDEX
    threads: THREADS
    log: "{results}/logs/{comparison}"
    shell:
        """
        {RMATS_CMD} \
          --b1 {input.b1} \
          --b2 {input.b2} \
          --gtf {params.gtf} \
          --od {params.od} \
          --readLength {params.read_len} \
          --nthread {threads} \
          --tmp {params.tmp} \
          {params.other_args} \
          {params.star_index} \

        """
# usage: rmats.py [options]
# optional arguments:
#   -h, --help            show this help message and exit
#   --version             show program's version number and exit
#   --gtf GTF             An annotation of genes and transcripts in GTF format
#   --b1 B1               A text file containing a comma separated list of the
#                         BAM files for sample_1. (Only if using BAM)
#   --b2 B2               A text file containing a comma separated list of the
#                         BAM files for sample_2. (Only if using BAM)
#   --s1 S1               A text file containing a comma separated list of the
#                         FASTQ files for sample_1. If using paired reads the
#                         format is ":" to separate pairs and "," to separate
#                         replicates. (Only if using fastq)
#   --s2 S2               A text file containing a comma separated list of the
#                         FASTQ files for sample_2. If using paired reads the
#                         format is ":" to separate pairs and "," to separate
#                         replicates. (Only if using fastq)
#   --od OD               The directory for final output from the post step
#   --tmp TMP             The directory for intermediate output such as ".rmats"
#                         files from the prep step
#   -t {paired,single}    Type of read used in the analysis: either "paired" for
#                         paired-end data or "single" for single-end data.
#                         Default: paired
#   --libType {fr-unstranded,fr-firststrand,fr-secondstrand}
#                         Library type. Use fr-firststrand or fr-secondstrand
#                         for strand-specific data. Default: fr-unstranded
#   --readLength READLENGTH
#                         The length of each read
#   --variable-read-length
#                         Allow reads with lengths that differ from --readLength
#                         to be processed. --readLength will still be used to
#                         determine IncFormLen and SkipFormLen
#   --anchorLength ANCHORLENGTH
#                         The anchor length. Default is 1
#   --tophatAnchor TOPHATANCHOR
#                         The "anchor length" or "overhang length" used in the
#                         aligner. At least "anchor length" NT must be mapped to
#                         each end of a given junction. The default is 6. (Only
#                         if using fastq)
#   --bi BINDEX           The directory name of the STAR binary indices (name of
#                         the directory that contains the SA file). (Only if
#                         using fastq)
#   --nthread NTHREAD     The number of threads. The optimal number of threads
#                         should be equal to the number of CPU cores. Default: 1
#   --tstat TSTAT         The number of threads for the statistical model. If
#                         not set then the value of --nthread is used
#   --cstat CSTAT         The cutoff splicing difference. The cutoff used in the
#                         null hypothesis test for differential splicing. The
#                         default is 0.0001 for 0.01% difference. Valid: 0 <=
#                         cutoff < 1. Does not apply to the paired stats model
#   --task {prep,post,both,inte,stat}
#                         Specify which step(s) of rMATS to run. Default: both.
#                         prep: preprocess BAMs and generate a .rmats file.
#                         post: load .rmats file(s) into memory, detect and
#                         count alternative splicing events, and calculate P
#                         value (if not --statoff). both: prep + post. inte
#                         (integrity): check that the BAM filenames recorded by
#                         the prep task(s) match the BAM filenames for the
#                         current command line. stat: run statistical test on
#                         existing output files
#   --statoff             Skip the statistical analysis
#   --paired-stats        Use the paired stats model
#   --novelSS             Enable detection of novel splice sites (unannotated
#                         splice sites). Default is no detection of novel splice
#                         sites
#   --mil MIL             Minimum Intron Length. Only impacts --novelSS
#                         behavior. Default: 50
#   --mel MEL             Maximum Exon Length. Only impacts --novelSS behavior.
#                         Default: 500
#   --allow-clipping      Allow alignments with soft or hard clipping to be used
#   --fixed-event-set FIXED_EVENT_SET
#                         A directory containing fromGTF.[AS].txt files to be
#                         used instead of detecting a new set of events
